framework/src/partitioner/NearestPositionsPartitioner.C

```diff
@@ -0,0 +1,218 @@
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NearestPositionsPartitioner.h"
+
+#include "MooseApp.h"
+#include "DelimitedFileReader.h"
+#include "KDTree.h"
+
+#include "libmesh/mesh_tools.h"
+#include "libmesh/elem.h"
+
+registerMooseObject("MooseApp", NearestPositionsPartitioner);
+
+InputParameters
+NearestPositionsPartitioner::validParams()
+{
+  InputParameters params = MoosePartitioner::validParams();
+
+  params.addParam<FileName>(
+      "positions_file",
+      "CSV file providing the positions coordinates (x,y,z) in the first three columns. Note that "
+      "Positions objects are not currently supported.");
+
+  params.addClassDescription("Create a partition of the positions, then assign mesh elements based "
+                             "on which position they are closest to.");
+
+  return params;
+}
+
+NearestPositionsPartitioner::NearestPositionsPartitioner(const InputParameters & params)
+  : MoosePartitioner(params), _mesh(*getCheckedPointerParam<MooseMesh *>("mesh"))
+{
+}
+
+NearestPositionsPartitioner::~NearestPositionsPartitioner() {}
+
+std::unique_ptr<Partitioner>
+NearestPositionsPartitioner::clone() const
+{
+  return _app.getFactory().clone(*this);
+}
+
+void
+NearestPositionsPartitioner::_do_partition(MeshBase & mesh, const unsigned int /*n*/)
+{
+  // Load the positions from the file
+  const FileName positions_file = getParam<FileName>("positions_file");
+  MooseUtils::DelimitedFileReader file(positions_file, &_communicator);
+  file.setFormatFlag(MooseUtils::DelimitedFileReader::FormatFlag::ROWS);
+  file.read();
+  const std::vector<Point> & positions_unindexed = file.getDataAsPoints();
+  const auto n_positions = positions_unindexed.size();
+
+  // Keep track of the index of each point, in case some positions are duplicated
+  std::vector<std::pair<unsigned int, Point>> positions(n_positions);
+  for (const auto i : index_range(positions))
+    positions[i] = std::make_pair(i, positions_unindexed[i]);
+
+  // Re-use this vector for neighbor searches
+  std::vector<std::size_t> return_index(2);
+
+  // Lloyd's algorithm:
+  // 1 - Build an initial close to even split
+  std::vector<std::vector<std::pair<unsigned int, Point>>> positions_array;
+  const auto n_bins = mesh.n_partitions();
+  positions_array.resize(n_bins);
+  for (auto & position_bin : positions_array)
+    // the number of bins has to be respected, so we may need to have a final bin with less
+    // positions
+    position_bin.reserve(ceil(n_positions / n_bins));
+  for (const auto i : index_range(positions))
+    positions_array[int(float(i) / n_positions * n_bins)].push_back(positions[i]);
+
+  if (n_bins > n_positions)
+    mooseError("Not implemented for more MPI processes than positions");
```

Comment:
probably a good use for the documented error?

Reply:
Removing this partitioner from the PR. The method worked fine up to a few dozen processes but completely broke after so it's not worth checking